File: swift-backend/Sources/App/Models/Goal.swift
@preconcurrency import Fluent

final class Goal: Model {
    static let schema = "goals"

    @ID(key: .id) var id: UUID?

    @Field(key: "name") var name: String

    @Field(key: "amount") var amount: Double

    @Field(key: "target_date") var targetDate: Date

    @Field(key: "user_id") var user: User

    @Field(key: "progress") var progress: Double

    @Timestamp(key: "created_at", on: .create) var createdAt: Date?

    @Timestamp(key: "updated_at", on: .update) var updatedAt: Date?

    @Timestamp(key: "deleted_at", on: .delete) var deletedAt: Date?

    init() {}

    init(
        id: UUID? = nil,
        name: String,
        amount: Double,
        targetDate: Date,
        userId: User.IDValue,
        progress: Double = 0
    ) {
        self.id = id
        self.name = name
        self.amount = amount
        self.targetDate = targetDate
        self.user.id = userId
        self.progress = progress
    }
}

File: api/src/api/models/Goal.scala
package api.models

import api.models.UserId

import java.time.Instant
import java.util.UUID
import io.circe.Encoder
import io.circe.Decoder

case class Goal(
    id: UUID,
    name: String,
    amount: Double,
    targetDate: Instant,
    userId: UserId,
    progress: Double,
    goalType: GoalType,
    createdAt: Instant,
    updatedAt: Instant,
    deletedAt: Option[Instant],
    assignedAccounts: List[GoalAccount]
)

enum GoalType(val value: String):
    case Retirement extends GoalType("retirement")
    case Custom extends GoalType("custom")

object GoalType:
    given Encoder[GoalType] = Encoder.encodeString.contramap[GoalType](_.value)

    given Decoder[GoalType] = Decoder
        .decodeString
        .emap { s =>
            GoalType.values.find(_.value == s) match
                case Some(goalType) =>
                    Right(goalType)
                case None =>
                    Left(s"Unknown GoalType: $s")
        }

    def fromString(s: String): Option[GoalType] = GoalType.values.find(_.value == s)
